File: app/src/main/java/com/wolkowiczmateusz/android/cleanmvpretrofitrxdaggerboilerplate/presentation/customviews/TriStateRecyclerView.java
package com.wolkowiczmateusz.android.cleanmvpretrofitrxdaggerboilerplate.presentation.customviews;

import android.content.Context;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.util.AttributeSet;
import android.util.Log;
import android.view.View;

public class TriStateRecyclerView extends RecyclerView {

    private View emptyView;
    private View loadingView;
    private AdapterDataObserver dataObserver = new AdapterDataObserver() {
        @Override
        public void onChanged() {
            super.onChanged();
            Adapter adapter = getAdapter();
            Log.i("RecycleError", this + "getAdapter() + adapter state" + adapter);
            boolean hasData = adapter.getItemCount() > 0;
            setVisibility(hasData ? VISIBLE : GONE);
            emptyView.setVisibility(hasData ? GONE : VISIBLE);
        }
    };

    public TriStateRecyclerView(Context context) {
        super(context);
    }

    public TriStateRecyclerView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    public TriStateRecyclerView(Context context, @Nullable AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
    }

    @Override
    public void setAdapter(Adapter adapter) {
        Log.i("RecycleError", this + "setAdapterBeforeSuper + adapter state" + adapter);
        super.setAdapter(adapter);
        loadingView.setVisibility(adapter == null ? VISIBLE : GONE);
        Log.i("RecycleError", this + "setAdapterAfterSuper + adapter state" + adapter);
        if (adapter != null) {
            adapter.registerAdapterDataObserver(dataObserver);
            dataObserver.onChanged();
        }
    }

    public void setEmptyView(View emptyView) {
        this.emptyView = emptyView;
    }

    public void setLoadingView(View loadingView) {
        this.loadingView = loadingView;
    }
}

File: app/src/main/java/com/wolkowiczmateusz/android/cleanmvpretrofitrxdaggerboilerplate/presentation/customviews/TriStateRecyclerView.kt
package com.wolkowiczmateusz.android.cleanmvpretrofitrxdaggerboilerplate.presentation.customviews

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.util.AttributeSet
import android.view.View

class TriStateRecyclerView : RecyclerView {

    private lateinit var emptyView: View
    private lateinit var loadingView: View
    private val dataObserver = object : RecyclerView.AdapterDataObserver() {
        override fun onChanged() {
            super.onChanged()
            val adapter = adapter
            val hasData = adapter!!.itemCount > 0
            visibility = if (hasData) View.VISIBLE else View.GONE
            emptyView.visibility = if (hasData) View.GONE else View.VISIBLE
        }
    }

    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)

    constructor(context: Context, attrs: AttributeSet?, defStyle: Int) : super(
        context,
        attrs,
        defStyle
    )

    override fun setAdapter(adapter: RecyclerView.Adapter<*>?) {
        super.setAdapter(adapter)
        loadingView.visibility = if (adapter == null) View.VISIBLE else View.GONE
        if (adapter != null) {
            adapter.registerAdapterDataObserver(dataObserver)
            dataObserver.onChanged()
        }
    }

    fun setEmptyView(emptyView: View) {
        this.emptyView = emptyView
    }

    fun setLoadingView(loadingView: View) {
        this.loadingView = loadingView
    }
}
